src/default-implementations.lisp

```diff
@@ -293,31 +293,161 @@ T on success.
+                                      (acquire-lock lock))
+                        (timeout ()
+                          nil)
+                        (:no-error ()
+                          (unwind-protect (progn ,@body)
+                            (release-lock lock))))
+                      (with-lock-held (lock)
+                        ,@body))))
+      (with-thunk ()
+        (if (>= (%semaphore-counter semaphore) 1)
+            (decf (%semaphore-counter semaphore))
+            (progn
+              ;; we need this loop because of a spurious wakeup possibility
+              (loop until (>= (%semaphore-counter semaphore) 1)
+                    do (cond
+                         ((null (condition-wait (%semaphore-condition-variable semaphore)
+                                                (%semaphore-lock semaphore)
+                                                :timeout timeout))
+                          (return-from wait-on-semaphore))
+                         ;; unfortunately cv-wait may return T on timeout too
+                         ((and deadline (>= (get-internal-real-time) deadline))
+                          (return-from wait-on-semaphore))
+                         (timeout
+                          (setf timeout (/ (- deadline (get-internal-real-time))
+                                           internal-time-units-per-second)))))
+              (decf (%semaphore-counter semaphore))))))))
 
 (defdfun semaphore-p (object)
   "Returns T if OBJECT is a semaphore; returns NIL otherwise."
   (typep object 'semaphore))
 
+;;; Mailboxes
+
+(defdfun make-mailbox (&key name initial-contents)
+  "Returns a new MAILBOX with messages in INITIAL-CONTENTS enqueued."
+  (check-type name (or null string))
+  (check-type initial-contents list)
+  (let ((locative (list nil)))
+    (make-%mailbox :name name
+                   :lock (make-lock name)
+                   :cvar (make-condition-variable :name name)
+                   :count (length initial-contents)
+                   :messages (append initial-contents locative)
+                   :messages-tail locative)))
+
+(defdfun mailbox-name (mailbox)
+  "Name of a MAILBOX. SETFable."
+  (%mailbox-name mailbox))
+
+(defdfun (setf mailbox-name) (name mailbox)
+  "Sets a MAILBOX name."
+  (setf (%mailbox-name mailbox) name))
+
+(defdfun list-mailbox-messages (mailbox)
+  "Returns a fresh list containing all the messages in the
+mailbox. Does not remove messages from the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (loop
+      for (i . j) on (%mailbox-messages mailbox)
+      until (null j) collect i)))
+
+(defdfun mailbox-count (mailbox)
+  "Returns the number of messages currently in the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (%mailbox-count mailbox)))
+
+(defdfun mailbox-empty-p (mailbox)
+  "Returns true if MAILBOX is currently empty, NIL otherwise."
+  (with-lock-held ((%mailbox-lock mailbox))
```

Comment:
Given that you take the lock here, there's the potential for needing a timeout as well, no?